how-to-update-to-php74-typed-property-without-bc-break/post.md

```diff
@@ -0,0 +1,171 @@
+
+[Typed Property](https://wiki.php.net/rfc/typed_properties_v2) is one of the PHP 7.4 feature that allow to write that previously like this:
+
+```php
+namespace Lib;
+
+class SomeClass
+{
+    /** @var int */
+    public $a;
+
+    /** @var string */
+    protected $b;
+
+    /** @var bool */
+    private $c;
+}
+```
+
+to this:
+
+```php
+namespace Lib;
+
+class SomeClass
+{
+    public int $a;
+
+    protected string $b;
+
+    private bool $c;
+}
+```
+
+If you follow [Semver](https://semver.org) for versioning, and you don't want to update to major change, eg: version 1 to version 2, changing this will make Break Backward Compatibility, for example:
+
+```php
+namespace Lib;
+
+class SomeClass
+{
+    protected string $b;
+}
+```
+
+has child in application consumer:
+
+```php
+namespace App;
+
+use Lib\SomeClass;
+
+class AChild extends SomeClass
+{
+    protected $b;
+}
+```
+
+will result a fatal error:
+
+```
+Fatal error: Type of AChild::$b must be string (as in class SomeClass)
+```
+
+see https://3v4l.org/X9Yvd . To avoid that, you should only change to private modifier only, so, the change will only to private property:
+
+```diff
+namespace Lib;
+
+class SomeClass
+{
+    /** @var int */
+    public $a;
+
+    /** @var string */
+    protected $b;
+
+-    /** @var bool */
+-    private $c;
++    private bool $c;
+}
+```
+
+Want to automate that? You can use [Rector](https://github.com/rectorphp/rector) for it. First, let say, we have a library with the following structure:
+
+```bash
+lib
+├── composer.json
+├── composer.lock
+├── src
+│   └── SomeClass.php
+```
+
+with composer.json config like this:
+
+```javascript
+{
+    "require": {
+        "php": "^7.4"
+    },
+    "autoload": {
+        "psr-4": {
+            "Lib\\": "src/"
+        }
+    }
+}
+
+```
+
+Now, what you need is require the rector as dev dependency:
+
+```bash
+cd lib/
+composer require --dev rector/rector
+```
+
+Then, create a `rector.php` configuration inside the `lib`:
+
+```php
+<?php
+
+declare(strict_types=1);
+
+use Rector\Core\Configuration\Option;
+use Rector\Core\ValueObject\PhpVersion;
+use Rector\Php74\Rector\Property\TypedPropertyRector;
+use Rector\Set\ValueObject\SetList;
+use Symfony\Component\DependencyInjection\Loader\Configurator\ContainerConfigurator;
+
+return static function (ContainerConfigurator $containerConfigurator): void {
+    $parameters = $containerConfigurator->parameters();
+    $parameters->set(Option::PATHS, [
+        __DIR__ . '/src'
+    ]);
+    $parameters->set(Option::PHP_VERSION_FEATURES, PhpVersion::PHP_74);
+
+    // import php 7.4 set list for php 7.4 features
+    $containerConfigurator->import(SetList::PHP_74);
+
+    // set Typed Property only for private property
+    $services = $containerConfigurator->services();
+    $services->set(TypedPropertyRector::class)
+        ->call('configure', [[
+            TypedPropertyRector::PRIVATE_PROPERTY_ONLY => true,
+        ]]);
```

Comment:
I wonder if this should be default, if you have to add it in every config here  :thinking: 

What do you think?

Reply:
I think currently is fine, which can work directly on projects that not redistributable

Reply:
I updated wording about TypedProperty can update all modifiers by default for project that not re-usable https://github.com/samsonasik/rector-blogposts/pull/2/commits/edf8588abde47bd97ad711dd7ea5c79f7e5375b3

Reply:
All right 👍